src/sage/crypto/key_exchange/diffie_hellman.py

```diff
@@ -0,0 +1,314 @@
+r"""
+Diffie-Hellman Public Key Exchange Scheme
+
+This module contains a toy implementation of the Diffie-Hellman public key exchange
+scheme.
+
+AUTHORS:
+
+- Vincent Macri (2024-07-15): initial version
+"""
+# ****************************************************************************
+#       Copyright (C) 2024 Vincent Macri <vincent.macri@ucalgary.ca>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#                  https://www.gnu.org/licenses/
+# ****************************************************************************
+
+from sage.misc.superseded import experimental
+
+from sage.crypto.key_exchange.key_exchange import KeyExchangeScheme
+
+from sage.arith.misc import is_prime
+from sage.misc.prandom import randint
+from sage.rings.integer import Integer
+from sage.rings.finite_rings.finite_field_constructor import GF
+from sage.rings.finite_rings.finite_field_prime_modn import FiniteField_prime_modn
+from sage.rings.finite_rings.integer_mod import IntegerMod_int
+from sage.structure.proof.proof import WithProof
+
+from typing import Union
+
+class DiffieHellman(KeyExchangeScheme):
+
+    @experimental(37305)
+    def __init__(self, p: Integer, g: Union[Integer, IntegerMod_int], proof: bool = True):
+        """
+        Create an instance of the Diffie-Hellman key exchange scheme using the
+        given prime ``p`` and base ``g``.
+
+        INPUT:
+
+        - ``p`` -- prime integer defining the field `\\GF{p}` that the key exchanges
```

Comment:
Should you specify the type of your parameters ? See the [Docstring INPUT block](https://doc.sagemath.org/html/en/developer/coding_basics.html#the-docstring-of-a-function-content) from the developer guide.

Reply:
Lots of code in Sage doesn't follow that (at least not in the format specified there). Since the type annotations show up in the documentation I figured that specifying types in the docstring would be unnecessary duplication. That part of the developer guide doesn't say anything about type annotation despite elsewhere in the developer guide saying that we should use type annotations when possible, so I had the impression that that specific part might be outdated.

Reply:
Personally I don't care.